File: src/org/irislang/jiris/irisclass/IrisArray.java
package org.irislang.jiris.irisclass;

import java.lang.reflect.Array;
import java.util.ArrayList;

import org.irislang.jiris.core.IrisClass;
import org.irislang.jiris.core.IrisContextEnvironment;
import org.irislang.jiris.core.IrisMethod;
import org.irislang.jiris.core.IrisModule;
import org.irislang.jiris.core.IrisThreadInfo;
import org.irislang.jiris.core.IrisValue;
import org.irislang.jiris.dev.IrisClassRoot;
import org.irislang.jiris.dev.IrisDevUtil;

public class IrisArray extends IrisClassRoot {

	public static IrisValue Initialize(IrisValue self, ArrayList<IrisValue> parameterList,
			ArrayList<IrisValue> variableParameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo) {
		
		if(variableParameterList != null) {
			ArrayList<IrisValue> arrayList = IrisDevUtil.GetNativeObjectRef(self);
			arrayList.addAll(variableParameterList);
		}
		
		return self;
	}
	
	public static IrisValue At(IrisValue self, ArrayList<IrisValue> parameterList,
			ArrayList<IrisValue> variableParameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo) {
		
		IrisValue index = parameterList.get(0);
		ArrayList<IrisValue> arrayList = IrisDevUtil.GetNativeObjectRef(self);
		
		if(!IrisDevUtil.CheckClass(index, "Integer")) {
			/* Error */
			return IrisDevUtil.Nil();
		}
		
		int indexNum = IrisDevUtil.GetInt(index);
		
		if(indexNum < 0) {
			return arrayList.get(arrayList.size() - (-indexNum % arrayList.size()));
		} else {
			if(indexNum > arrayList.size()) {
				for(int i = 0; i < indexNum - arrayList.size(); ++i) {
					arrayList.add(IrisDevUtil.Nil());
				}
				
				return IrisDevUtil.Nil();
			}
			else {
				return arrayList.get(indexNum);
			}
		}
	}
	
	public static IrisValue Set(IrisValue self, ArrayList<IrisValue> parameterList,
			ArrayList<IrisValue> variableParameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo) {
		IrisValue index = parameterList.get(0);
		IrisValue targetValue = parameterList.get(1);
		ArrayList<IrisValue> arrayList = IrisDevUtil.GetNativeObjectRef(self);
		if(!IrisDevUtil.CheckClass(index, "Integer")) {
			/* Error */
			return IrisDevUtil.Nil();
		}

		int indexNum = IrisDevUtil.GetInt(index);

		if(indexNum < 0) {
			arrayList.set(arrayList.size() - (-indexNum % arrayList.size()), targetValue);
		} else {
			if(indexNum >= arrayList.size()) {
				for(int i = 0; i < indexNum - arrayList.size(); ++i) {
					arrayList.add(IrisDevUtil.Nil());
				}
				arrayList.add(targetValue);
			}
			else {
				arrayList.set(indexNum, targetValue);
			}
		}
		return self;
	}

	public static IrisValue Push(IrisValue self, ArrayList<IrisValue> parameterList,
			ArrayList<IrisValue> variableParameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo) {
		IrisValue targetValue = parameterList.get(0);
		ArrayList<IrisValue> arrayList = IrisDevUtil.GetNativeObjectRef(self);
		arrayList.add(targetValue);
		return self;
	}
	
	public static IrisValue Pop(IrisValue self, ArrayList<IrisValue> parameterList,
			ArrayList<IrisValue> variableParameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo) {
		ArrayList<IrisValue> arrayList = IrisDevUtil.GetNativeObjectRef(self);
		return arrayList.remove(arrayList.size() - 1);
	}

	public static IrisValue Size(IrisValue self, ArrayList<IrisValue> parameterList,
			ArrayList<IrisValue> variableParameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo) {
		ArrayList<IrisValue> arrayList = IrisDevUtil.GetNativeObjectRef(self);
		return IrisDevUtil.CreateInt(arrayList.size());
	}

	public static IrisValue GetIterator(IrisValue self, ArrayList<IrisValue> parameterList, ArrayList<IrisValue> variableParameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo) {
		ArrayList<IrisValue> params = new ArrayList<IrisValue>(1);
		params.add(0, self);
		// ** Error **
		try {
			return IrisDevUtil.CreateInstance(IrisDevUtil.GetClass("ArrayIterator"), params, context, threadInfo);
		} catch (Throwable throwable) {
			throwable.printStackTrace();
			return IrisDevUtil.Nil();
		}
	}
	
	@Override
	public String NativeClassNameDefine() {
		return "Array";
	}

	@Override
	public IrisClass NativeSuperClassDefine() {
		return IrisDevUtil.GetClass("Object");
	}

	@Override
	public IrisModule NativeUpperModuleDefine() {
		return null;
	}

	@Override
	public Object NativeAlloc() {
		return new ArrayList<IrisValue>();
	}

	@Override
	public void NativeClassDefine(IrisClass classObj) throws Throwable {
		classObj.AddInstanceMethod(IrisArray.class, "Initialize", "__format", 0, true, IrisMethod.MethodAuthority.Everyone);
		classObj.AddInstanceMethod(IrisArray.class, "At", "[]", 1, false, IrisMethod.MethodAuthority.Everyone);
		classObj.AddInstanceMethod(IrisArray.class, "Set", "[]=", 2, false, IrisMethod.MethodAuthority.Everyone);
		
		classObj.AddInstanceMethod(IrisArray.class, "Push", "push", 1, false, IrisMethod.MethodAuthority.Everyone);
		classObj.AddInstanceMethod(IrisArray.class, "Pop", "pop", 0, false, IrisMethod.MethodAuthority.Everyone);
		classObj.AddInstanceMethod(IrisArray.class, "GetIterator", "get_iterator", 0, false, IrisMethod
				.MethodAuthority.Everyone);
	}

}

File: src/main/java/org/irislang/jiris/irisclass/IrisArray.kt
package org.irislang.jiris.irisclass

import java.lang.reflect.Array
import java.util.ArrayList

import org.irislang.jiris.core.IrisClass
import org.irislang.jiris.core.IrisContextEnvironment
import org.irislang.jiris.core.IrisMethod
import org.irislang.jiris.core.IrisModule
import org.irislang.jiris.core.IrisThreadInfo
import org.irislang.jiris.core.IrisValue
import org.irislang.jiris.core.exceptions.IrisExceptionBase
import org.irislang.jiris.core.exceptions.fatal.IrisTypeNotCorretException
import org.irislang.jiris.dev.IrisClassRoot
import org.irislang.jiris.dev.IrisDevUtil

class IrisArray : IrisClassRoot() {

    override fun NativeClassNameDefine(): String {
        return "Array"
    }

    override fun NativeSuperClassDefine(): IrisClass {
        return IrisDevUtil.GetClass("Object")!!
    }

    override fun NativeUpperModuleDefine(): IrisModule? {
        return null
    }

    override fun NativeAlloc(): Any {
        return ArrayList<IrisValue>()
    }

    @Throws(IrisExceptionBase::class)
    override fun NativeClassDefine(classObj: IrisClass) {
        classObj.AddInstanceMethod(IrisArray::class.java, "Initialize", "__format", 0, true, IrisMethod.MethodAuthority.Everyone)
        classObj.AddInstanceMethod(IrisArray::class.java, "At", "[]", 1, false, IrisMethod.MethodAuthority.Everyone)
        classObj.AddInstanceMethod(IrisArray::class.java, "Set", "[]=", 2, false, IrisMethod.MethodAuthority.Everyone)

        classObj.AddInstanceMethod(IrisArray::class.java, "Push", "push", 1, false, IrisMethod.MethodAuthority.Everyone)
        classObj.AddInstanceMethod(IrisArray::class.java, "Pop", "pop", 0, false, IrisMethod.MethodAuthority.Everyone)
        classObj.AddInstanceMethod(IrisArray::class.java, "GetIterator", "get_iterator", 0, false, IrisMethod
                .MethodAuthority.Everyone)
    }

    companion object {

        @JvmStatic
        fun Initialize(self: IrisValue, parameterList: ArrayList<IrisValue>,
                       variableParameterList: ArrayList<IrisValue>, context: IrisContextEnvironment, threadInfo: IrisThreadInfo): IrisValue {

            val arrayList = IrisDevUtil.GetNativeObjectRef<ArrayList<IrisValue>>(self)
            arrayList.addAll(variableParameterList)

            return self
        }

        @Throws(IrisExceptionBase::class)
        @JvmStatic
        fun At(self: IrisValue, parameterList: ArrayList<IrisValue>,
               variableParameterList: ArrayList<IrisValue>, context: IrisContextEnvironment, threadInfo: IrisThreadInfo): IrisValue {

            val index = parameterList[0]
            val arrayList = IrisDevUtil.GetNativeObjectRef<ArrayList<IrisValue>>(self)

            if (!IrisDevUtil.CheckClass(index, "Integer")) {
                /* Error */
                throw IrisTypeNotCorretException(threadInfo.currentFileName, threadInfo.currentLineNumber,
                        "Parameter of [] can only be an Integer.")
            }

            val indexNum = IrisDevUtil.GetInt(index)

            if (indexNum < 0) {
                return arrayList[arrayList.size - -indexNum % arrayList.size]
            } else {
                if (indexNum > arrayList.size) {
                    for (i in 0 until indexNum - arrayList.size) {
                        arrayList.add(IrisDevUtil.Nil())
                    }
                    return IrisDevUtil.Nil()
                } else {
                    return arrayList[indexNum]
                }
            }
        }

        @Throws(IrisExceptionBase::class)
        @JvmStatic
        fun Set(self: IrisValue, parameterList: ArrayList<IrisValue>,
                variableParameterList: ArrayList<IrisValue>, context: IrisContextEnvironment, threadInfo: IrisThreadInfo): IrisValue {
            val index = parameterList[0]
            val targetValue = parameterList[1]
            val arrayList = IrisDevUtil.GetNativeObjectRef<ArrayList<IrisValue>>(self)
            if (!IrisDevUtil.CheckClass(index, "Integer")) {
                /* Error */
                throw IrisTypeNotCorretException(threadInfo.currentFileName, threadInfo.currentLineNumber,
                        "Index parameter of []= can only be an Integer.")
            }

            val indexNum = IrisDevUtil.GetInt(index)

            if (indexNum < 0) {
                arrayList[arrayList.size - -indexNum % arrayList.size] = targetValue
            } else {
                if (indexNum >= arrayList.size) {
                    for (i in 0 until indexNum - arrayList.size) {
                        arrayList.add(IrisDevUtil.Nil())
                    }
                    arrayList.add(targetValue)
                } else {
                    arrayList[indexNum] = targetValue
                }
            }
            return self
        }

        @JvmStatic
        fun Push(self: IrisValue, parameterList: ArrayList<IrisValue>,
                 variableParameterList: ArrayList<IrisValue>, context: IrisContextEnvironment, threadInfo: IrisThreadInfo): IrisValue {
            val targetValue = parameterList[0]
            val arrayList = IrisDevUtil.GetNativeObjectRef<ArrayList<IrisValue>>(self)
            arrayList.add(targetValue)
            return self
        }

        @JvmStatic
        fun Pop(self: IrisValue, parameterList: ArrayList<IrisValue>,
                variableParameterList: ArrayList<IrisValue>, context: IrisContextEnvironment, threadInfo: IrisThreadInfo): IrisValue {
            val arrayList = IrisDevUtil.GetNativeObjectRef<ArrayList<IrisValue>>(self)
            return arrayList.removeAt(arrayList.size - 1)
        }

        @JvmStatic
        fun Size(self: IrisValue, parameterList: ArrayList<IrisValue>,
                 variableParameterList: ArrayList<IrisValue>, context: IrisContextEnvironment, threadInfo: IrisThreadInfo): IrisValue {
            val arrayList = IrisDevUtil.GetNativeObjectRef<ArrayList<IrisValue>>(self)
            return IrisDevUtil.CreateInt(arrayList.size)
        }

        @Throws(IrisExceptionBase::class)
        @JvmStatic
        fun GetIterator(self: IrisValue, parameterList: ArrayList<IrisValue>, variableParameterList: ArrayList<IrisValue>, context: IrisContextEnvironment, threadInfo: IrisThreadInfo): IrisValue {
            val params = ArrayList<IrisValue>(1)
            params.add(0, self)
            // ** Error **
            return IrisDevUtil.CreateInstance(IrisDevUtil.GetClass("ArrayIterator")!!, params, context, threadInfo)
        }
    }

}
